File: typescript/src/problems101_200/reorder-list.ts
import SinglyListNode from '../concepts/singly-list-node';

class Problem {
  reorderList(head: SinglyListNode | undefined) {
    throw new Error(`Not implemented (${head}).`);
  }
}

class Default extends Problem {
  reorderList(head: SinglyListNode | undefined) {
    const mid: SinglyListNode = this.getMid(head);
    let right: SinglyListNode = this.reverse(mid);
    mid.next = undefined;
    let left: SinglyListNode = head;

    let current: SinglyListNode = new SinglyListNode();
    let flag: boolean = true;
    while (left && right) {
      if (flag) {
        current.next = left;
        left = left.next;
      } else {
        current.next = right;
        right = right.next;
      }
      current = current.next;
      flag = !flag;
    }
  }

  reverse(node: SinglyListNode | undefined): SinglyListNode | undefined {
    let previous: SinglyListNode | undefined = undefined;
    let current: SinglyListNode | undefined = node;
    while (current) {
      let temp = current.next;
      current.next = previous;
      previous = current;
      current = temp;
    }
    return previous;
  }

  getMid(node: SinglyListNode | undefined): SinglyListNode | undefined {
    let slow = node;
    let fast = node;
    while (fast && fast.next) {
      slow = slow.next;
      fast = fast.next.next;
    }
    return slow;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems101_200/reorder_list.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.singly_list_node import SinglyListNode


class Problem(ABC):
    @abstractmethod
    def reorder_list(self, head: SinglyListNode | None) -> None:
        pass


class Default(Problem):
    @override
    def reorder_list(self, head: SinglyListNode | None) -> None:
        mid: SinglyListNode | None = self.get_mid(head)
        right: SinglyListNode | None = self.reverse(mid)
        mid.next = None
        left: SinglyListNode | None = head

        current: SinglyListNode = SinglyListNode()
        flag: bool = True
        while left and right:
            if flag:
                current.next = left
                left = left.next
            else:
                current.next = right
                right = right.next
            current = current.next
            flag = not flag

    def reverse(self, node: SinglyListNode | None) -> SinglyListNode | None:
        previous: SinglyListNode | None = None
        current: SinglyListNode | None = node
        while current:
            temp = current.next
            current.next = previous
            previous = current
            current = temp
        return previous

    def get_mid(self, node: SinglyListNode | None) -> SinglyListNode | None:
        slow: SinglyListNode = node
        fast: SinglyListNode | None = node
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow


SOLUTIONS: list[Problem] = [Default()]
